File: graph.cpp
#include<iostream>
#include<queue>
#define INF 10000
#define WHITE 0
#define GRAY 1
#define BLACK 2

using namespace std;
int color[5];

int findmin(int *val,int *mstset)
{
int i;
int min_val=INF;
for(i=0;i<5;i++)
{
if((mstset[i]==0)&&val[i]<=min_val)
	min_val=i;
}
return min_val;
}

int printMST(int parent[], int n, int g[5][5])
{
int i;
   for (i = 1; i < 5; i++)
     cout<<parent[i]<<"-"<<i<<" "<< g[i][parent[i]]<<endl;
}

void PRIMS_SPANNING_TREE(int g[5][5])
{
int parent[5];
int val[5];
int mstset[5];
int i,v;

for(i=0;i<5;i++)
{
	val[i]=INF;
	mstset[i]=0;
	parent[i]=0;
}
parent[0]=-1;
val[0]=0;
for(i=0;i<5;i++)
{
	int u=findmin(val,mstset);
	mstset[u]=1;
	for(v=0;v<5;v++)
	{	
		if((g[u][v]!=0)&&(mstset[v]==0)&&g[u][v]<val[v])
			{
				val[v]=g[u][v];	
				parent[v]=u;
			}
	}
}

printMST(parent,5,g);
}

void DFS_VISIT(int g[5][5],int );
void DFS(int g[5][5])
{
int i;
for(i=0;i<5;i++)
	color[i]=WHITE;
for(i=0;i<5;i++)
{
	if(color[i]==WHITE)
		DFS_VISIT(g,i);
}
}
void DFS_VISIT(int g[5][5],int i)
{
color[i]=GRAY;

cout<<i<<"\t";
int k;

for(k=0;k<5;k++)
{
if(g[i][k]&&color[k]==WHITE)
	DFS_VISIT(g,k);
}
color[i]=BLACK;
}


void BFS(int g[5][5])
{
int source,i,j,temp;
queue<int>q;
for(i=0;i<5;i++)
	color[i]=WHITE;

cout<<"enter source vertex ";
cin>>source;
q.push(source);
color[source]=GRAY;
while(!q.empty())
{
	temp=q.front();
	for(i=0;i<5;i++)
	{
		if(g[temp][i])
		{
			if(color[i]==WHITE)
			{
				q.push(i);
				color[i]=GRAY;
			}
		}
	}
	cout<<temp<<"\t";
	color[temp]=BLACK;
	q.pop();
}

}

void create(int g[][5],int n)
{
	int i,j,temp,data,wt;
	for(i=0;i<n;i++)
	{
		for(j=0;j<n;j++)
		g[i][j]=0;
	}

	for(i=0;i<n;i++)
	{
		cout<<"enter number of vertices to node "<<i<<"\t";
		cin>>temp;
		{
			for(j=0;j<temp;j++)
			{
			cout<<"enter data && wight "<<j<<endl;
			cin>>data;
			cin>>wt;
			g[i][data]=wt;
			}
		}
	}

	for(i=0;i<n;i++)
	{
		for(j=0;j<n;j++)
		{
			cout<<g[i][j]<<"\t";
		}
		cout<<endl;
	}
}

int main()
{
int n;

int g[5][5];
int i,j;
for(i=0;i<5;i++)
{
	for(j=0;j<5;j++)
		g[i][j]=0;
}
//create(g,n);

g[0][1]=2;
g[0][3]=6;
g[1][0]=2;
g[1][2]=3;
g[1][3]=8;
g[1][4]=5;
g[2][1]=3;
g[2][4]=7;
g[3][0]=6;
g[3][1]=8;
g[3][4]=9;
g[4][1]=5;
g[4][2]=7;
g[4][3]=9;
cout<<"\n------------BFS traversal is----------\n";
BFS(g);
cout<<"\n----------DFS traversal is----------\n";
DFS(g);
cout<<"\n calling prims\n";
PRIMS_SPANNING_TREE(g);
return 0;
}

File: Graph.java
import java.io.*;
import java.util.*;

class Graph
{
	class Node
	{
		Node ptr;
		int data;
		int wight;
		Node(int data,int wight)
		{
			this.data=data;
			this.ptr=null;		
			this.wight=wight;
		}
	}
	final static int WHITE=0;
	final static int GRAY=1;
	final static int BLACK=2;
	Node arr[];
	int color[];

	Graph(int n)
	{
		arr=new Node[n+1];
		color=new int[n+1];
		for(int i=1;i<=n;i++)
		{
			arr[i]=null;
			color[i]=WHITE;;
		}
	}

	void add(int data,int i,int w)
	{
		arr[i]=insert(arr[i],data,w);
	}

	Node insert(Node root, int data,int wight)
	{
		Node temp=new Node(data,wight);
		temp.ptr=null;
		if(root==null)
		{
			root=temp;
		}
		else
		{
			Node walk=root;
			while(walk.ptr!=null)
				walk=walk.ptr;
			walk.ptr=temp;
		}
		return root;
			
	}

	void show_graph(int n)
	{
		int i;
		Node walk;
		for(i=1;i<=n;i++)
		{
			walk=arr[i];
			System.out.print(i+"--> ");
			while(walk!=null)
			{
		    		System.out.print(walk.data+"\t"+walk.wight);
		    		walk=walk.ptr;
			}
	                System.out.println();
		}
	}

	void BFS(int source)
	{
		Queue<Integer> q=new LinkedList<>();
		q.add(source);
		Node temp;
		int val1,val2;
		while(!q.isEmpty())
		{
			val1=q.remove();
			temp=arr[val1];
			while(temp!=null)
			{
				val2=temp.data;
				if(color[val2]==WHITE)
				{
					q.add(val2);
					color[val2]=GRAY;
				}
				temp=temp.ptr;
			}
			color[val1]=BLACK;
			System.out.println("data is "+val1);	
		}
	}
	void DFS(int source,int n)
	{
		int i;
		for(i=1;i<=n;i++)
		{
			if(color[i]==WHITE)
				DFS_visit(i);
		}
	}
	void DFS_visit(int i)
	{
		color[i]=GRAY;
		Node temp;
		int val;
		temp=arr[i];
		while(temp!=null)
		{
			val=temp.data;
			if(color[val]==WHITE)
			{
				DFS_visit(val);
			}
			temp=temp.ptr;
		}
		System.out.println(i+"\t");
		color[i]=BLACK;
	}
	int doesContainsCycle()
	{
		int flag=0;
		return flag;
	}

	public static void main(String args[])throws Exception
	{
		Scanner sc=new Scanner(System.in);
		int n,i,j,adjacent,data,wt;
		System.out.println("Enter number of vertices in the graph ");
		n=sc.nextInt();
		Graph g=new Graph(n);
		for(i=1;i<=n;i++)
		{
			System.out.println("how many adjacent nodes for node "+i+"\t");
			adjacent=sc.nextInt();
			for(j=1;j<=adjacent;j++)
			{
				System.out.println("enter node "+j+"\t");
				data=sc.nextInt();
				System.out.println("enter wt of edge "+i+"  to  "+data+"\t");
				wt=sc.nextInt();
				g.add(i,data,wt);      
			}
		}
		g.show_graph(n); 
//		g.BFS(1);
//		g.DFS(1,n);
	}
}
